server/add_service.go

```diff
@@ -0,0 +1,74 @@
+package server
+
+import (
+	"encoding/json"
+	"io"
+
+	"golang.org/x/net/context"
+)
+
+// Add is a function that takes two ints and returns an int.
+type Add func(context.Context, int, int) int
+
+// PureAdd is a pure service and can ignore the context.
+func PureAdd(_ context.Context, a, b int) int { return a + b }
+
+// AddViaResource is a non-pure service and must use the context.
+func AddViaResource(r Resource) func(context.Context, int, int) int {
+	return func(ctx context.Context, a, b int) int {
+		return PureAdd(ctx, r.GetA(ctx, a), b)
+	}
+}
+
+// Resource TODO
+type Resource struct{}
+
+// GetA TODO
+func (r Resource) GetA(ctx context.Context, a int) int {
+	return a
+}
+
+// AddRequest can be autogenerated.
```

Comment:
I think they could be generated, but I haven't yet sussed out the details. I'm hopeful that `go generate` is up to the task.